texar/tf/utils/dynamic_decode.py

```diff
@@ -0,0 +1,331 @@
+            if maximum_iterations.get_shape().ndims != 0:
+                raise ValueError("maximum_iterations must be a scalar")
+
+    initial_finished, initial_inputs, initial_state = decoder.initialize()
+
+    zero_outputs = _create_zero_outputs(decoder.output_size,
+                                        decoder.output_dtype,
+                                        decoder.batch_size)
+
+    if maximum_iterations is not None:
+        initial_finished = tf.logical_or(
+            initial_finished, 0 >= maximum_iterations)
+    initial_sequence_lengths = tf.zeros_like(
+        initial_finished, dtype=tf.int32)
+    initial_time = tf.constant(0, dtype=tf.int32)
+
+    def _shape(batch_size, from_shape):
+        if (not isinstance(from_shape, tensor_shape.TensorShape) or
+                from_shape.ndims == 0):
+            return None
+        else:
+            batch_size = tf.get_static_value(
+                tf.convert_to_tensor(
+                    batch_size, name="batch_size"))
+            return tensor_shape.TensorShape([batch_size]).\
+                concatenate(from_shape)
+
+    dynamic_size = True
+
+    def _create_ta(s, d):
+        return tf.TensorArray(
+            dtype=d,
+            size=0 if dynamic_size else maximum_iterations,
+            dynamic_size=dynamic_size,
+            element_shape=_shape(decoder.batch_size, s))
+
+    initial_outputs_ta = nest.map_structure(_create_ta, decoder.output_size,
+                                            decoder.output_dtype)
+
+    def condition(unused_time, unused_outputs_ta, unused_state, unused_inputs,
+                  finished, unused_sequence_lengths):
+        cond = tf.logical_not(tf.reduce_all(finished))
+        cond_time = (maximum_iterations is None or
+                     unused_time < maximum_iterations)
+        ret = tf.logical_and(cond, tf.convert_to_tensor(cond_time))
+        return ret
+
+    def body(time, outputs_ta, state, inputs, finished, sequence_lengths):
+        r"""Internal while_loop body.
+
+        Args:
+            time: scalar int32 tensor.
+            outputs_ta: structure of TensorArray.
+            state: (structure of) state tensors and TensorArrays.
+            inputs: (structure of) input tensors.
+            finished: bool tensor (keeping track of what's finished).
+            sequence_lengths: int32 tensor (keeping track of time of finish).
+
+        Returns:
+            `(time + 1, outputs_ta, next_state, next_inputs, next_finished,
+            next_sequence_lengths)`.
+        """
+        (next_outputs, state) = decoder.step(time, inputs, state)
+        reach_max = tf.equal(time+1, maximum_iterations)
+        (decoder_finished, next_inputs, decoder_state) = tf.cond(
+            reach_max,
+            lambda: (tf.cast(tf.ones_like(finished), tf.bool),
+                     inputs, state),
+            lambda: decoder.next_inputs(time, next_outputs, state)
+        )
+        if decoder.tracks_own_finished:
```

Comment:
Add comment here that this is what we have modified from TF original dynamic_decode